service/history/chasm/engine.go

```diff
@@ -0,0 +1,241 @@
+package chasm
+
+import (
+	"context"
+)
+
+// now that is not exported, a struct is good enough
+// instead of an interface
+type engine interface {
+	newInstance(
+		context.Context,
+		EntityKey,
+		func(MutableContext) (Component, error),
+		...TransitionOption,
+	) (ComponentRef, error)
+	updateWithNewInstance(
+		context.Context,
+		EntityKey,
+		func(MutableContext) (Component, error),
+		func(MutableContext, Component) error,
+		...TransitionOption,
+	) (ComponentRef, error)
+
+	updateComponent(
+		context.Context,
+		ComponentRef,
+		func(MutableContext, Component) error,
+		...TransitionOption,
+	) (ComponentRef, error)
+	readComponent(
+		context.Context,
+		ComponentRef,
+		func(Context, Component) error,
+		...TransitionOption,
+	) error
+
+	pollComponent(
+		context.Context,
+		ComponentRef,
+		func(Context, Component) bool,
+		func(MutableContext, Component) error,
+		...TransitionOption,
+	) (ComponentRef, error)
+}
+
+type BusinessIDReusePolicy int
+
+const (
+	BusinessIDReusePolicyAllowDuplicate BusinessIDReusePolicy = iota
+	BusinessIDReusePolicyRejectDuplicate
+)
+
+type BusinessIDConflictPolicy int
+
+const (
+	BusinessIDConflictPolicyFail BusinessIDConflictPolicy = iota
+	BusinessIDConflictPolicyTermiateExisting
+	BusinessIDConflictPolicyUseExisting
+)
+
+type transitionOptions struct {
+}
+
+type TransitionOption func(*transitionOptions)
+
+// (only) this transition will not be persisted
+// The next non-speculative transition will persist this transition as well.
+// Compared to the EntityEphemeral() operation on RegistrableComponent,
+// the scope of this operation is limited to a certain transition,
+// while the EntityEphemeral() applies to all transitions.
+// TODO: we need to figure out a way to run the tasks
+// generated in a speculative transition
+func WithSpeculative() TransitionOption {
+	panic("not implemented")
+}
+
+// this only applies to NewEntity and UpdateWithNewEntity
+func WithBusinessIDPolicy(
+	reusePolicy BusinessIDReusePolicy,
+	conflictPolicy BusinessIDConflictPolicy,
+) TransitionOption {
+	panic("not implemented")
+}
+
+// Not needed for V1
```

Comment:
True, this isn't needed but if we were to do it, there's a type safe way using a closure that gets an instance and returns field references.